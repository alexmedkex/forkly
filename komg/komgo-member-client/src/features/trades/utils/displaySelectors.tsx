import { CreditRequirements, TRADE_SCHEMA, ITrade, Commodity, Status } from '@komgo/types'
import { List, Map } from 'immutable'
import { get } from 'lodash'
import Numeral from 'numeral'
import * as React from 'react'
import { Label as SemanticUILabel, Popup, SemanticCOLORS } from 'semantic-ui-react'
import { capitalize, sentenceCase } from '../../../utils/casings'
import { displayDate, isDate } from '../../../utils/date'
import { Counterparty } from '../../counterparties/store/types'
import { Roles } from '../../letter-of-credit-legacy/constants/roles'
import { ILetterOfCredit } from '../../letter-of-credit-legacy/types/ILetterOfCredit'
import { IMember } from '../../members/store/types'
import {
  CREDIT_REQUIREMENT_DISPLAY_VALUES,
  LOC_CATEGORY_COLOURS,
  NOTENOUGHINFO,
  READABLE_LOC_STATUS,
  TradingRole,
  UNRECOGNISED_STATUS,
  TRADE_STATUS
} from '../constants'
import { ITradeEnriched } from '../store/types'
import { findFieldFromSchema } from '../../../store/common/selectors/displaySelectors'
import { Strings } from '../../receivable-discounting-legacy/resources/strings'

export const displayYesOrNo = (trueOrFalse: boolean): string => (trueOrFalse ? 'Yes' : 'No')

export const displayPercentage = (percent: number = 0): string => {
  return Numeral(percent / 100).format('% 0.00')
}

export const displayQuantity = (quantity?: number, priceUnit?: string, nullValueDisplay = NOTENOUGHINFO): string => {
  if (!quantity || !priceUnit) {
    return nullValueDisplay
  }
  return `${formatQuantity(quantity)} ${priceUnit}`
}

export const formatQuantity = (quantity: number): string => {
  return `${Numeral(quantity).format('0,0')}`
}

export const displayTolerance = displayPercentage

export const displayCommodity = (commodity: string): string => {
  if (commodity === undefined) {
    // TODO: Remove check once properly implemented from VAKT (MM)
    return 'BFOET'
  }
  return Object.values(Commodity).includes(commodity) ? sentenceCase(commodity) : commodity
}

export const displayPrice = (
  price?: number,
  currency?: string,
  priceUnit?: string,
  nullValueDisplay = NOTENOUGHINFO
): string => {
  if (!price || !currency || !priceUnit) {
    return nullValueDisplay
  }
  return `${formatPrice(price)} ${currency}/${priceUnit}`
}

export const formatPrice = (price: number): string => {
  return `${Numeral(price).format('0,0.00')}`
}

export const displayEtrmId = (trade: ITradeEnriched, companyStaticId?: string, role?: string): string => {
  if (trade.buyer === companyStaticId || role === Roles.ISSUING_BANK) {
    return `${trade.buyerEtrmId}`
  } else if (trade.seller === companyStaticId || role === Roles.ADVISING_BANK) {
    return `${trade.sellerEtrmId}`
  }
  return NOTENOUGHINFO
}

export const getHumanReadableLCStatus = (status: string): string => {
  return (
    Object.keys(READABLE_LOC_STATUS).find(
      readableStatus =>
        (READABLE_LOC_STATUS as any)[readableStatus].find((matchingStatus: string) => matchingStatus === status) !==
        undefined
    ) || UNRECOGNISED_STATUS
  )
}

export const displayTradeStatus = (tradeStatus: string): string =>
  tradeStatus === TRADE_STATUS.ToBeDiscounted ? Strings.SellerTradeStatusDefault : sentenceCase(tradeStatus)

export const displayLCStatus = (status?: string): JSX.Element | undefined => {
  if (!status) {
    return
  }

  const humanReadableLCStatus = getHumanReadableLCStatus(status)
  return makeLCStatusPopup(humanReadableLCStatus, (LOC_CATEGORY_COLOURS as any)[
    humanReadableLCStatus
  ] as SemanticCOLORS)
}

const makeLCStatusPopup = (text: string, color?: SemanticCOLORS): JSX.Element => {
  return (
    <Popup color={color} trigger={<SemanticUILabel circular={true} color={color} basic={!color && true} />}>
      {capitalize(text.toLocaleLowerCase().replace(/_/g, ' '))}
    </Popup>
  )
}

export const selectVisibleTrades = (trades: Map<string, any>, tradeIds: List<string>): ITrade[] => {
  return tradeIds.toJS().map((id: string) => {
    const trade = trades.get(id)
    if (!trade) {
      throw new Error(`Unexpected trade: ${id}`)
    }
    return trade.toJS()
  })
}

export const findCommonNameByStaticId = (parties: Array<Counterparty | IMember>, staticId?: string) => {
  const member = parties.find((c: Counterparty | IMember) => c.staticId === staticId)
  return member ? get(member, 'x500Name.CN') : `${staticId} not found`
}

export const addBuyerSellerEnrichedData = (
  currentCompanyStaticId: string,
  trades: ITrade[],
  members: Counterparty[] | IMember[]
): ITradeEnriched[] => {
  return trades.map((trade: ITradeEnriched) => ({
    ...trade,
    sellerName: findCommonNameByStaticId(members, trade.seller),
    buyerName: findCommonNameByStaticId(members, trade.buyer),
    tradingRole: currentCompanyStaticId === trade.seller ? TradingRole.SELLER : TradingRole.BUYER
  }))
}

export const selectCounterparty = (trade: ITradeEnriched, tradingRole: TradingRole) => {
  return tradingRole === TradingRole.BUYER ? trade.sellerName : trade.buyerName
}

export const displayCreditRequirement = (creditRequirement?: string): string => {
  const creditRequirementHumanReadableValue =
    CREDIT_REQUIREMENT_DISPLAY_VALUES[creditRequirement ? creditRequirement : '']
  return creditRequirementHumanReadableValue ? creditRequirementHumanReadableValue[0] : UNRECOGNISED_STATUS
}

export const selectNewestLetterOfCreditWithTradeId = (letters: ILetterOfCredit[], tradeId?: string) => {
  return letters
    .filter(l => (l.tradeAndCargoSnapshot ? l.tradeAndCargoSnapshot.trade._id === tradeId : false))
    .sort(
      (a, b) =>
        a.updatedAt && b.updatedAt && typeof a.updatedAt === 'string' && typeof b.updatedAt === 'string'
          ? Date.parse(b.updatedAt) - Date.parse(a.updatedAt)
          : -1
    )[0]
}

export const findFieldFromTradeSchema = (field: string, name: string, schema: any = TRADE_SCHEMA) => {
  return findFieldFromSchema(field, name, schema)
}

export const displayObject = (obj, property, options): string => {
  return JSON.stringify(
    obj,
    (key, value) => {
      if (key) {
        return displayValue(value, key)
      }
      return value
    },
    2
  )
    .replace(/{|}|,|"/g, '')
    .replace(/\\n/g, '\n  ')
    .replace('  \n', '')

  /* output example
    _id: 5c3f0a4d6d25550e4d744bc9
    id: idparcel2
    laycanPeriod:
      startDate: 2018/09/01
      endDate: 2018/09/30
    vesselIMO: 1
    vesselName: Andrej
    loadingPort: Banja luka
    dischargeArea: Sarajevo
    inspector: Boom
    deemedBLDate: 2018/09/01
    quantity: 10
  */
}

export const displayValue = (value, property, options = {}): string => {
  switch (property) {
    case 'maxTolerance':
    case 'minTolerance':
      return displayTolerance(value)
    case 'commodity':
      return displayCommodity(value)
    case 'price':
    case 'amount':
      return formatPrice(value)
    case 'quantity':
      return formatQuantity(value)
    default:
      return handleDefault(value, property, options)
  }
}

const handleDefault = (value, property, options) => {
  if (typeof value === 'boolean') {
    return value ? 'true' : 'false'
  }
  if (isDate(value)) {
    return displayDate(value)
  }
  if (value && typeof value === 'object') {
    return displayObject(value, property, options)
  }
  return value ? value : '-'
}
