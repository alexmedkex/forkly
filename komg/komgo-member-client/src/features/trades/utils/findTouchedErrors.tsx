import { FormikErrors, FormikTouched } from 'formik'
import _ from 'lodash'

export const findTouchedErrors = (errors: FormikErrors<any>, touched: FormikTouched<any>): FormikErrors<any> => {
  const touchedErrors = {}
  Object.keys(errors).forEach(field => {
    if (field.includes('.')) {
      const arrayField = field.split('.')
      let fieldTouched: any = touched
      arrayField.forEach(item => {
        if (fieldTouched[item]) {
          fieldTouched = fieldTouched[item]
        }
      })
      if (fieldTouched === true) {
        touchedErrors[field] = errors[field]
      }
    } else if (touched[field]) {
      touchedErrors[field] = errors[field]
    }
  })
  return touchedErrors
}

export const findFormikErrors = (errors: FormikErrors<any>, touched: FormikTouched<any>): FormikErrors<any> => {
  const touchedErrors = {}
  Object.keys(errors).forEach(field => {
    if (_.get(touched, field)) {
      touchedErrors[field] = errors[field]
    }
  })
  return touchedErrors
}
