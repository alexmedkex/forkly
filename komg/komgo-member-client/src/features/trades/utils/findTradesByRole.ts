import { ITradeEnriched } from '../store/types'
import { TradingRole } from '../constants'

export const findTradesByRole = (trades: ITradeEnriched[], tradingRole: TradingRole, company: string) =>
  trades.filter(trade => trade[tradingRole] === company)
