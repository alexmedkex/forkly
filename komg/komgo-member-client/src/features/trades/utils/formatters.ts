import { displayDate, dateFormats } from '../../../utils/date'
import { ITradeEnriched, ICreateOrUpdateTrade, ITradeDocument } from '../store/types'
import {
  initialCargoData,
  initialTradeData,
  initialParcelData,
  TRADING_ROLE_OPTIONS,
  tradeDefaultEmptyValues
} from '../constants'
import {
  Law,
  ICargo,
  ICargoBase,
  IParcel,
  ITrade,
  Commodity,
  PaymentTermsEventBase,
  DeliveryTerms,
  PriceOption,
  PaymentTermsOption,
  ModeOfTransport,
  TRADE_SCHEMA_VERSION,
  CARGO_SCHEMA_VERSION,
  PARCEL_SCHEMA_VERSION,
  CreditRequirements,
  PaymentTermsTimeUnit,
  PaymentTermsDayType,
  PaymentTermsWhen
} from '@komgo/types'
import { Document } from '../../document-management'
import { PERMITTED_MIME_TYPES } from '../../document-management/utils/permittedMimeTypes'
import { IParcelWithOverrides } from '../components/cargo-form-fields/ParcelData'
import { v4 } from 'uuid'
import _ from 'lodash'

// Format for dates are different on api and on form
export const formatEditTrade = (
  trade: ITradeEnriched,
  cargos: ICargo[],
  documents: Document[]
): ICreateOrUpdateTrade => {
  const cargo = cargos[0] ? formatCargoDates(cargos[0]) : ({ ...initialCargoData, cargoId: v4() } as ICargo)
  const formattedTrade = formatTradeDate(trade)

  const { lawOther, commodityOther, eventBaseOther, deliveryTermsOther } = getOtherFields(formattedTrade)

  const { paymentTerms, paymentTermsOption } = getPaymentTermsFields(formattedTrade, eventBaseOther)

  let tradeData: ITradeEnriched = {
    ...initialTradeData,
    ...formattedTrade,
    status: trade.status,
    law: lawOther ? Law.Other : trade.law,
    commodity: commodityOther ? Commodity.Other : trade.commodity,
    paymentTerms,
    deliveryTerms: deliveryTermsOther ? DeliveryTerms.Other : trade.deliveryTerms,
    paymentTermsOption,
    priceFormula: trade.priceFormula ? trade.priceFormula : '',
    version: trade.version ? trade.version : TRADE_SCHEMA_VERSION.V2
  }

  const requiredDocuments = tradeData.requiredDocuments || []

  tradeData = _.merge({}, tradeDefaultEmptyValues, removeUndefinedOrNull(tradeData))

  return {
    trade: {
      ...tradeData,
      requiredDocuments
    },
    cargo: {
      ...initialCargoData,
      ...cargo,
      parcels: cargo.parcels.map(formatParcelForEdit),
      version: cargo.version ? cargo.version : CARGO_SCHEMA_VERSION.V2
    },
    lawOther,
    commodityOther,
    eventBaseOther,
    deliveryTermsOther,
    documents: formatDocuments(documents)
  }
}

const getPaymentTermsFields = (trade: ITradeEnriched, eventBaseOther?: string) => {
  const paymentTermsOption = trade.paymentTermsOption ? trade.paymentTermsOption : null

  const paymentTerms =
    paymentTermsOption === PaymentTermsOption.Deferred
      ? {
          ...trade.paymentTerms,
          eventBase: eventBaseOther ? PaymentTermsEventBase.Other : trade.paymentTerms.eventBase
        }
      : initialTradeData.paymentTerms

  return { paymentTermsOption, paymentTerms }
}

const getOtherFields = (trade: ITradeEnriched) => {
  const lawOther = !Object.values(Law).includes(trade.law) && !!trade.law ? trade.law : ''
  const commodityOther = !Object.values(Commodity).includes(trade.commodity) && !!trade.commodity ? trade.commodity : ''
  const eventBaseOther =
    trade.paymentTerms &&
    !Object.values(PaymentTermsEventBase).includes(trade.paymentTerms.eventBase) &&
    !!trade.paymentTerms.eventBase
      ? trade.paymentTerms.eventBase
      : ''
  const deliveryTermsOther =
    !Object.values(DeliveryTerms).includes(trade.deliveryTerms) && !!trade.deliveryTerms ? trade.deliveryTerms : ''

  return {
    lawOther,
    commodityOther,
    deliveryTermsOther,
    eventBaseOther
  }
}

export const formatParcelForEdit = (parcel: IParcel): IParcelWithOverrides => {
  const modeOfTransportOther =
    !Object.values(ModeOfTransport).includes(parcel.modeOfTransport) && !!parcel.modeOfTransport
      ? parcel.modeOfTransport
      : ''

  return {
    ...initialParcelData,
    ...parcel,
    modeOfTransport: modeOfTransportOther ? ModeOfTransport.Other : parcel.modeOfTransport,
    modeOfTransportOther,
    version: parcel.version ? parcel.version : PARCEL_SCHEMA_VERSION.V2
  }
}

export const formatCargoData = (cargo: ICargo, sourceId: string): ICargoBase => ({
  ...cargo,
  sourceId
})

export const formatCargoDates = (cargo: ICargo): ICargo => ({
  ...cargo,
  parcels: cargo.parcels.map(parcel => ({
    ...parcel,
    laycanPeriod: {
      startDate: displayDate(parcel.laycanPeriod.startDate, dateFormats.inputs),
      endDate: displayDate(parcel.laycanPeriod.endDate, dateFormats.inputs)
    },
    deemedBLDate: displayDate(parcel.deemedBLDate, dateFormats.inputs)
  }))
})

export const formatTradeDate = (trade: ITradeEnriched): ITradeEnriched => ({
  ...(trade as ITrade),
  deliveryPeriod: {
    startDate: trade.deliveryPeriod ? displayDate(trade.deliveryPeriod.startDate, dateFormats.inputs) : '',
    endDate: trade.deliveryPeriod ? displayDate(trade.deliveryPeriod.endDate, dateFormats.inputs) : ''
  },
  dealDate: displayDate(trade.dealDate, dateFormats.inputs),
  contractDate: displayDate(trade.contractDate, dateFormats.inputs)
})

export const findMimeType = (documentName: string): string => {
  const extension = documentName.substr(documentName.lastIndexOf('.') + 1)
  const [type] = Object.values(PERMITTED_MIME_TYPES).filter(t => t.includes(extension))
  return type || ''
}

export const formatDocuments = (documents: Document[]): ITradeDocument[] =>
  documents.map(document => ({
    id: document.id,
    name: document.name.substr(0, document.name.lastIndexOf('.')),
    categoryId: document.category.id,
    typeId: document.type.id,
    fileName: document.name,
    fileType: findMimeType(document.name),
    file: null
  }))

export const generateInitialFormData = (
  trade: ITradeEnriched,
  tradeMovements: ICargo[],
  documents: Document[],
  tradingRole: string,
  company: string
): ICreateOrUpdateTrade => {
  if (!tradingRole) {
    return null
  }

  const setCompanyNameBasedOnTradingRole =
    tradingRole === TRADING_ROLE_OPTIONS.BUYER ? { buyer: company } : { seller: company }

  if (trade) {
    const editTrade = formatEditTrade(trade, tradeMovements, documents)

    return {
      ...editTrade,
      trade: {
        ...editTrade.trade,
        ...setCompanyNameBasedOnTradingRole
      }
    }
  }

  return {
    documents: [],
    lawOther: '',
    commodityOther: '',
    eventBaseOther: '',
    deliveryTermsOther: '',
    trade: {
      ...initialTradeData,
      dealDate: displayDate(new Date(), dateFormats.inputs),
      ...setCompanyNameBasedOnTradingRole,
      creditRequirement:
        tradingRole === TRADING_ROLE_OPTIONS.BUYER
          ? CreditRequirements.StandbyLetterOfCredit
          : (CreditRequirements.OpenCredit as any)
    } as any,
    cargo: { ...initialCargoData, cargoId: v4() } as any // initialize cargoId in order to provide uniqueness. Temporary solution
  }
}

export const handlePaymentOptionsChange = (value: PaymentTermsOption, trade: ITrade) => {
  trade.paymentTerms =
    value === PaymentTermsOption.Deferred
      ? {
          time: '' as any,
          timeUnit: PaymentTermsTimeUnit.Days,
          dayType: PaymentTermsDayType.Calendar,
          when: PaymentTermsWhen.After,
          eventBase: PaymentTermsEventBase.BL
        }
      : undefined
}

export const replaceEmptyStringsAndNullWithUndefined = (values: any) => {
  const entries = Object.entries(values).map(([key, value]) => [
    value === '' || value === null ? undefined : value,
    key
  ])
  const cleaned = entries.reduce(
    (memo: any, [value, key]: any) => ({
      ...memo,
      [key]: _.isArray(value)
        ? value.map(v => replaceEmptyStringsAndNullWithUndefined(v))
        : _.isObject(value)
          ? replaceEmptyStringsAndNullWithUndefined(value)
          : value
    }),
    {}
  )
  return cleaned
}

export const removeUndefinedOrNull = <T extends object = object>(obj: T): T =>
  Object.keys(obj)
    .filter(key => obj[key] != null)
    .reduce<T>(
      (memo: object, key: string) =>
        ({
          ...memo,
          [key]: _.isArray(obj[key])
            ? obj[key].map(v => removeUndefinedOrNull(v))
            : _.isObject(obj[key])
              ? removeUndefinedOrNull(obj[key])
              : obj[key]
        } as T),
      {} as any
    )
