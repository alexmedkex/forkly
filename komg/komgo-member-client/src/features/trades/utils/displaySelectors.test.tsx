import { ITradeEnriched } from '../store/types'
import { Map, List, fromJS } from 'immutable'
import {
  selectVisibleTrades,
  displayQuantity,
  displayPrice,
  displayLCStatus,
  selectCounterparty,
  getHumanReadableLCStatus,
  displayCreditRequirement,
  displayEtrmId,
  selectNewestLetterOfCreditWithTradeId,
  findFieldFromTradeSchema,
  displayValue,
  displayYesOrNo
} from './displaySelectors'
import { capitalize } from '../../../utils/casings'

import {
  NOTENOUGHINFO,
  UNRECOGNISED_STATUS,
  LOC_STATUS,
  LOC_CATEGORIES,
  TradingRole,
  CREDIT_REQUIREMENT_LABEL
} from '../constants'
import { fakeTrade, fakeParcel } from '../../letter-of-credit-legacy/utils/faker'
import { Roles } from '../../letter-of-credit-legacy/constants/roles'
import { TradeSource, ITrade, CreditRequirements } from '@komgo/types'

const exampleCompany = 'my oil company'
const exampleCompanyId = '123'

const exampleTrades: ITradeEnriched[] = [
  {
    ...fakeTrade(),
    source: TradeSource.Vakt,
    status: 'OK',
    sourceId: '1',
    _id: 'a',
    buyer: exampleCompanyId,
    buyerName: exampleCompany,
    seller: '321',
    sellerName: exampleCompany
  },
  {
    ...fakeTrade(),
    source: TradeSource.Komgo,
    status: 'LOST',
    sourceId: '2',
    _id: 'b',
    seller: exampleCompanyId,
    sellerName: exampleCompany,
    buyer: 'texaco',
    buyerName: '098'
  },
  {
    ...fakeTrade(),
    source: TradeSource.Vakt,
    status: 'SHIPPED',
    sourceId: '2',
    _id: 'c',
    seller: exampleCompanyId,
    sellerName: exampleCompany,
    buyer: 'texaco',
    buyerName: '098'
  }
]

describe('selectVisibleTrades', () => {
  let trades: Map<string, ITrade>
  let tradeIds: List<string>
  beforeAll(() => {
    const map: any = exampleTrades.reduce(
      (memo, trade: ITrade) => ({
        ...memo,
        [trade._id!]: trade
      }),
      {}
    )
    trades = fromJS(map)
  })
  beforeEach(() => {
    tradeIds = List<string>()
  })
  it('selects no trades if the tradeId list is empty', () => {
    const selectedTrades = selectVisibleTrades(trades, tradeIds)

    expect(selectedTrades.length).toBe(0)
  })
  it('selects the correct trades if the tradeId list is populated', () => {
    tradeIds = tradeIds.push(exampleTrades[0]._id!).push(exampleTrades[2]._id!)

    const selectedTrades = selectVisibleTrades(trades, tradeIds)

    expect(selectedTrades.length).toEqual(2)
    expect(selectedTrades[0]).toEqual(exampleTrades[0])
    expect(selectedTrades[1]).toEqual(exampleTrades[2])
  })
  it('throws error if an ID is given which is not in the trades store', () => {
    tradeIds = tradeIds.push('abc')
    expect(() => selectVisibleTrades(trades, tradeIds)).toThrow()
  })
})

describe('displayYesOrNo', () => {
  it('displays Yes when true', () => {
    expect(displayYesOrNo(true)).toBe('Yes')
  })
  it('displays No when false', () => {
    expect(displayYesOrNo(false)).toBe('No')
  })
})

describe('displayQuantity', () => {
  it('displays quantity and price unit if both given', () => {
    const result = displayQuantity(1, 'bbl')
    expect(result).toEqual('1 bbl')
  })
  it('displays nothing if no quantity information given', () => {
    const result = displayQuantity(undefined, 'bbl')
    expect(result).toEqual(NOTENOUGHINFO)
  })
  it('displays nothing if no price unit information given', () => {
    const result = displayQuantity(1)
    expect(result).toEqual(NOTENOUGHINFO)
  })
  it('displays comma separated quantity', () => {
    const result = displayQuantity(40000, 'bbl')
    expect(result).toEqual('40,000 bbl')
  })
})

describe('displayPrice', () => {
  it('displays full information if given', () => {
    expect(displayPrice(1, 'usd', 'bbl')).toEqual('1.00 usd/bbl')
  })
  it('displays nothing if no price given', () => {
    expect(displayPrice(undefined, 'usd', 'bbl')).toEqual(NOTENOUGHINFO)
  })
  it('displays nothing if no currency given', () => {
    expect(displayPrice(1, undefined, 'bbl')).toEqual(NOTENOUGHINFO)
  })
  it('displays nothing if no priceunit given', () => {
    expect(displayPrice(1, 'usd', undefined)).toEqual(NOTENOUGHINFO)
  })
  it('displays comma separated quantity with 2 decimal points', () => {
    const result = displayPrice(40000, 'usd', 'bbl')
    expect(result).toEqual('40,000.00 usd/bbl')
  })
})

describe('displayLCStatus', () => {
  it('displays nothing if there is an undefined status', () => {
    expect(displayLCStatus()).toBeUndefined()
  })
  it('displays a red label if the trade is waiting to be financed', () => {
    const jsxResult: JSX.Element = displayLCStatus(LOC_STATUS.TO_BE_FINANCED)!

    expect(jsxResult.props.trigger.props.color).toEqual('red')
    expect(jsxResult.props.color).toEqual('red')
  })
  it('contains "Waiting to be financed" text if the trade is waiting to be financed', () => {
    const jsxResult: JSX.Element = displayLCStatus(LOC_STATUS.TO_BE_FINANCED)!

    expect(jsxResult.props.children).toContain(capitalize((LOC_CATEGORIES as any)[LOC_STATUS.TO_BE_FINANCED]))
  })
  it('displays a green label if the LC is of issued type', () => {
    const jsxResult: JSX.Element = displayLCStatus(LOC_STATUS.ADVISED)!

    expect(jsxResult.props.trigger.props.color).toEqual('green')
  })
  it('contains "LC issued" text if the LC is of issued type', () => {
    const jsxResult: JSX.Element = displayLCStatus(LOC_STATUS.RECEIVED)!

    expect(jsxResult.props.children).toEqual(capitalize(LOC_CATEGORIES.ISSUED))
  })
  it('displays an uncoloured label if the LC is refused', () => {
    const jsxResult: JSX.Element = displayLCStatus(LOC_STATUS.REFUSED)!

    expect(jsxResult.props.trigger.props.color).toEqual(undefined)
    expect(jsxResult.props.trigger.props.basic).toEqual(true)
  })
  it('contains "LC closed" text if the LC is closed', () => {
    const jsxResult: JSX.Element = displayLCStatus(LOC_STATUS.REFUSED)!

    expect(jsxResult.props.children).toContain(capitalize(LOC_CATEGORIES.CLOSED))
  })
  it('displays an uncoloured basic label if the LC is paid', () => {
    const jsxResult: JSX.Element = displayLCStatus(LOC_STATUS.PAID)!

    expect(jsxResult.props.trigger.props.color).toEqual(undefined)
    expect(jsxResult.props.trigger.props.basic).toEqual(true)
  })
  it('contains "LC closed" text if the LC is paid', () => {
    const jsxResult: JSX.Element = displayLCStatus(LOC_STATUS.PAID)!

    expect(jsxResult.props.children).toContain(capitalize(LOC_CATEGORIES.CLOSED))
  })
  it('displays a yellow label if the LC status is submitted', () => {
    const jsxResult: JSX.Element = displayLCStatus(LOC_STATUS.SUBMITTED)!

    expect(jsxResult.props.trigger.props.color).toEqual('yellow')
  })
})

describe('selectCounterparty', () => {
  it('selects the seller id if we are the buyer', () => {
    const result = selectCounterparty(exampleTrades[0], TradingRole.BUYER)
    expect(result).toEqual(exampleTrades[0].sellerName)
  })
  it('selects the buyer id if we are the seller', () => {
    const result = selectCounterparty(exampleTrades[0], TradingRole.SELLER)
    expect(result).toEqual(exampleTrades[0].buyerName)
  })
})

describe('getHumanReadableLCStatus', () => {
  it('selects to be financed for a TO_BE_FINANCED status', () => {
    const result = getHumanReadableLCStatus(LOC_STATUS.TO_BE_FINANCED)

    expect(result).toEqual(LOC_CATEGORIES.TO_BE_FINANCED)
  })
  it('selects LC issued for a ISSUED category status', () => {
    const result = getHumanReadableLCStatus(LOC_STATUS.ADVISED)

    expect(result).toEqual(LOC_CATEGORIES.ISSUED)
  })
  it('selects LC in progress for SUBMITTED status', () => {
    const result = getHumanReadableLCStatus(LOC_STATUS.SUBMITTED)

    expect(result).toEqual(LOC_CATEGORIES.IN_PROGRESS)
  })
  it('selects LC closed for REFUSED status', () => {
    const result = getHumanReadableLCStatus(LOC_STATUS.REFUSED)

    expect(result).toEqual(LOC_CATEGORIES.CLOSED)
  })
  it('selects Status unrecognized for a status not matching', () => {
    const result = getHumanReadableLCStatus('rasdfjh34kjk')

    expect(result).toEqual(UNRECOGNISED_STATUS)
  })
})

describe('displayCreditRequirement', () => {
  it('selects Documentary LC for a DOCUMENTARY_LETTER_OF_CREDIT creditRequirement', () => {
    const result = displayCreditRequirement(CreditRequirements.DocumentaryLetterOfCredit)

    expect(result).toEqual(CREDIT_REQUIREMENT_LABEL.DOCUMENTARY_LETTER_OF_CREDIT)
  })
  it('selects Standby LC for a STANDBY_LETTER_OF_CREDIT creditRequirement', () => {
    const result = displayCreditRequirement(CreditRequirements.StandbyLetterOfCredit)

    expect(result).toEqual(CREDIT_REQUIREMENT_LABEL.STANDBY_LETTER_OF_CREDIT)
  })
  it('selects Open Credit for a OPEN_CREDIT creditRequirement', () => {
    const result = displayCreditRequirement(CreditRequirements.OpenCredit)

    expect(result).toEqual(CREDIT_REQUIREMENT_LABEL.OPEN_CREDIT)
  })
  it('selects Offset for a OFFSET creditRequirement', () => {
    const result = displayCreditRequirement(CreditRequirements.Offset)

    expect(result).toEqual(CREDIT_REQUIREMENT_LABEL.OFFSET)
  })
  it('selects Status unrecognized for a status not matching', () => {
    const result = displayCreditRequirement('ewefwefwefwef')

    expect(result).toEqual(UNRECOGNISED_STATUS)
  })
})

describe('displayEtrmId', () => {
  it('Displays buyerEtrmId if company is buyer and role is empty', () => {
    const trade = exampleTrades[0]
    const result = displayEtrmId(trade, exampleCompanyId)

    expect(result).toEqual(trade.buyerEtrmId)
  })
  it('Displays sellerEtrmId if company is seller and role is empty', () => {
    const trade = exampleTrades[1]
    const result = displayEtrmId(trade, exampleCompanyId)

    expect(result).toEqual(trade.sellerEtrmId)
  })
  it('Display buyerEtrmId if company is buyer and credit requirement is open credit', () => {
    const trade = { ...exampleTrades[0], creditRequirement: CreditRequirements.OpenCredit }
    const result = displayEtrmId(trade, exampleCompanyId)

    expect(result).toEqual(trade.buyerEtrmId)
  })
  it('Display sellerEtrmId if company is seller and credit requirement is open credit', () => {
    const trade = { ...exampleTrades[1], creditRequirement: CreditRequirements.OpenCredit }
    const result = displayEtrmId(trade, exampleCompanyId)

    expect(result).toEqual(trade.sellerEtrmId)
  })
  it('Display buyerEtrmId if company is empty and role is issuing bank', () => {
    const trade = exampleTrades[0]
    const result = displayEtrmId(trade, undefined, Roles.ISSUING_BANK)

    expect(result).toEqual(trade.buyerEtrmId)
  })
  it('Displays sellerEtrmId if company is empty and role is advising bank', () => {
    const trade = exampleTrades[0]
    const result = displayEtrmId(trade, undefined, Roles.ADVISING_BANK)

    expect(result).toEqual(trade.sellerEtrmId)
  })
})

describe('selectNewestLetterOfCreditWithTradeId', () => {
  const exampleLetters: any = [
    {
      _id: '123',
      status: 'request rejected',
      tradeAndCargoSnapshot: { trade: { _id: '123_1' } },
      updatedAt: '2018-12-06T11:44:39.704Z'
    },
    {
      _id: '234',
      status: 'request rejected',
      updatedAt: '2014-05-24T11:44:39.123Z',
      tradeAndCargoSnapshot: { trade: { _id: '234_1' } }
    },
    {
      _id: '234',
      status: 'requested',
      updatedAt: '2016-11-04T13:12:33.321Z',
      tradeAndCargoSnapshot: { trade: { _id: '234_1' } }
    },
    {
      _id: '234',
      status: 'nope',
      tradeAndCargoSnapshot: { trade: { _id: '234_1' } }
    }
  ]
  it('returns a letter of credit matching the id given', () => {
    const letter = selectNewestLetterOfCreditWithTradeId(exampleLetters, '123_1')

    expect(letter).not.toBeUndefined()
    expect(letter!._id).toEqual('123')
  })
  it('returns undefined if no letter of credit is matched', () => {
    expect(selectNewestLetterOfCreditWithTradeId(exampleLetters, 'noMatch')).toBeUndefined()
  })
  it('returns the newest letter if two letters of credit match', () => {
    expect(selectNewestLetterOfCreditWithTradeId(exampleLetters, '234_1')!.status).toEqual('requested')
  })
})

describe('findFieldFromSchema', () => {
  it('should find field name from schema', () => {
    expect(findFieldFromTradeSchema('title', 'sellerEtrmId')).toBe("Seller's reference")
    expect(findFieldFromTradeSchema('title', 'paymentTerms.eventBase')).toBe('Eventbase')
  })
  it('if field name do not exist should return same text which is passed as field name', () => {
    expect(findFieldFromTradeSchema('title', 'test')).toBe('test')
    expect(findFieldFromTradeSchema('title', 'test.test')).toBe('test.test')
  })
})

describe('displayValue', () => {
  describe('displays', () => {
    it('maxTolerance', () => {
      expect(displayValue(1, 'maxTolerance')).toEqual('1.00%')
    })

    it('minTolerance', () => {
      expect(displayValue(1, 'minTolerance')).toEqual('1.00%')
    })

    it('default commodity', () => {
      expect(displayValue(undefined, 'commodity')).toEqual('BFOET')
    })

    it('a date', () => {
      expect(displayValue(new Date('2019-02-15'), 'a date')).toEqual('2019/02/15')
    })

    it('a parcel', () => {
      expect(displayValue(fakeParcel(), 'parcel')).toMatchSnapshot()
    })

    it('a price', () => {
      expect(displayValue(3800, 'price')).toEqual('3,800.00')
    })

    it('a quantity', () => {
      expect(displayValue(2000, 'quantity')).toEqual('2,000')
    })

    it('a sellerEtrmId', () => {
      expect(displayValue('trade-90', 'sellerEtrmId')).toEqual('trade-90')
    })

    it('a null value', () => {
      expect(displayValue(null, 'whatever')).toEqual('-')
    })

    it('a undefined value', () => {
      expect(displayValue(undefined, 'whatever')).toEqual('-')
    })
  })
})
