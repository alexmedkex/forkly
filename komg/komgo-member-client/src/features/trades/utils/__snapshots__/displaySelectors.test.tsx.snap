// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displayValue displays a parcel 1`] = `
"
  _id: 123
  id: 1234
  laycanPeriod: 
    startDate: 2019/01/03
    endDate: 2019/01/03
  modeOfTransport: VESSEL
  vesselIMO: 1
  vesselName: Test
  loadingPort: Test
  dischargeArea: Test
  inspector: Test
  deemedBLDate: 2019/01/03
  quantity: 1
"
`;
